### PR TITLE
Convert string types to Enums in Struct py:enum gen code

### PR DESCRIPTION
**Pull request by roshanjrajan-zip**
When serializing and deserializing Enums for Structs, the current code expects Enum member variables to be a string instead of an Enum. What this means is that the thrift file shows that an Enum is passed in the struct, when actually a string is passed in. This is very confusing can arbitrarily lead to KeyError in the deserialization logic.

However, changing this behavior is potentially breaking. To avoid breaking existing users, check if the type is an Enum when setting the member variable and convert the type to an Enum if it is a string. This allows the actual type to be an Enum and simplifies the serialization/deserialization logic to use Enums instead of converting strings.

Given a thrift file like the following:
```python
enum TestEnum {
  TestEnum0 = 0,
  TestEnum1 = 1,
}

struct TestStruct {
    1: optional string param1
    2: optional TestEnum param2
}
```

this is how you would create a struct before and after the change but both are valid.
```python
# Before
TestStruct(param1="test_string", param2=TestEnum.TestEnum1.name) # or pass in "TestEnum1"

# After
TestStruct(param1="test_string", param2=TestEnum.TestEnum1)
```

Serialization/Deserialization Code Change Example- 
```python
class TestStruct(object):
    """
    Attributes:
     - param1
     - param2

    """


    def __init__(self, param1=None, param2=None,):
        self.param1 = param1
        self.param2 = param2

++    def __setattr__(self, name, value):
++		if name == "param2":
++	        super().__setattr__(name, value if hasattr(value, 'value') else TestEnum.__members__.get(value))
++			return
++        super().__setattr__(name, value)

    def read(self, iprot):
        if iprot._fast_decode is not None and isinstance(iprot.trans, TTransport.CReadableTransport) and self.thrift_spec is not None:
            iprot._fast_decode(self, iprot, [self.__class__, self.thrift_spec])
            return
        iprot.readStructBegin()
        while True:
            (fname, ftype, fid) = iprot.readFieldBegin()
            if ftype == TType.STOP:
                break
            if fid == 1:
                if ftype == TType.STRING:
                    self.param1 = iprot.readString().decode('utf-8', errors='replace') if sys.version_info[0] == 2 else iprot.readString()
                else:
                    iprot.skip(ftype)
            elif fid == 2:
                if ftype == TType.I32:
--                    self.param2 = TestEnum(iprot.readI32()).name
++                    self.param2 = TestEnum(iprot.readI32())
                else:
                    iprot.skip(ftype)
            else:
                iprot.skip(ftype)
            iprot.readFieldEnd()
        iprot.readStructEnd()

    def write(self, oprot):
        if oprot._fast_encode is not None and self.thrift_spec is not None:
            oprot.trans.write(oprot._fast_encode(self, [self.__class__, self.thrift_spec]))
            return
        oprot.writeStructBegin('TestStruct')
        if self.param1 is not None:
            oprot.writeFieldBegin('param1', TType.STRING, 1)
            oprot.writeString(self.param1.encode('utf-8') if sys.version_info[0] == 2 else self.param1)
            oprot.writeFieldEnd()
        if self.param2 is not None:
            oprot.writeFieldBegin('param2', TType.I32, 2)
--            oprot.writeI32(TestEnum[self.param2].value)
++           oprot.writeI32(self.param2.value)
            oprot.writeFieldEnd()
        oprot.writeFieldStop()
        oprot.writeStructEnd()

    def validate(self):
        return

    def __repr__(self):
        L = ['%s=%r' % (key, value)
             for key, value in self.__dict__.items()]
        return '%s(%s)' % (self.__class__.__name__, ', '.join(L))

    def __eq__(self, other):
        return isinstance(other, self.__class__) and self.__dict__ == other.__dict__

    def __ne__(self, other):
        return not (self == other)
all_structs.append(TestStruct)
TestStruct.thrift_spec = (
    None,  # 0
    (1, TType.STRING, 'param1', 'UTF8', None, ),  # 1
    (2, TType.I32, 'param2', None, None, ),  # 2
)
fix_spec(all_structs)
del all_structs
```

  

- [ ] Did you create an [Apache Jira](https://issues.apache.org/jira/projects/THRIFT/issues/) ticket?  ([Request account here](https://selfserve.apache.org/jira-account.html), not required for trivial changes)
- [ ] If a ticket exists: Does your pull request title follow the pattern "THRIFT-NNNN: describe my issue"?
- [x] Did you squash your changes to a single commit?  (not required, but preferred)
- [x] Did you do your best to avoid breaking changes?  If one was needed, did you label the Jira ticket with "Breaking-Change"?
- [ ] If your change does not involve any code, include `[skip ci]` anywhere in the commit message to free up build resources.

### CODE REVIEW

File: compiler/cpp/src/thrift/generate/t_py_generator.cc [hunk 1]
```diff
@@ -905,6 +905,32 @@ void t_py_generator::generate_py_struct_definition(ostream& out,
...
+        has_enum = true;
+        break;
+      }
+    }
+
+    if (has_enum) {
+      out << endl;
+      indent(out) << "def __setattr__(self, name, value):" << endl;
+      indent_up();
+      for (m_iter = members.begin(); m_iter != members.end(); ++m_iter) {
```
Comment (fishy) [inline compiler/cpp/src/thrift/generate/t_py_generator.cc:922]: can you add a test (and also paste the generated python code into PR discussion) on when a struct has 2 fields that's enum (and they are different enum types)?

Comment (roshanjrajan-zip) [reply compiler/cpp/src/thrift/generate/t_py_generator.cc:922]: Updated the test! Here is the entire generated code file
```python
#
# Autogenerated by Thrift Compiler (0.19.0)
#
# DO NOT EDIT UNLESS YOU ARE SURE THAT YOU KNOW WHAT YOU ARE DOING
#
#  options string: py:enum
#

from thrift.Thrift import TType, TMessageType, TFrozenDict, TException, TApplicationException
from thrift.protocol.TProtocol import TProtocolException
from thrift.TRecursive import fix_spec
from enum import IntEnum

import sys
import shared_types.ttypes

from thrift.transport import TTransport
all_structs = []


class TestEnum(IntEnum):
    TestEnum0 = 0
    TestEnum1 = 1



class TestStruct(object):
    """
    Attributes:
     - param1
     - param2
     - param3

    """


    def __init__(self, param1=None, param2=None, param3=None,):
        self.param1 = param1
        self.param2 = param2
        self.param3 = param3

    def __setattr__(self, name, value):
        if name == "param2":
            super().__setattr__(name, value if hasattr(value, 'value') else TestEnum.__members__.get(value))
            return
        if name == "param3":
            super().__setattr__(name, value if hasattr(value, 'value') else shared_types.ttypes.SharedEnum.__members__.get(value))
            return
        super().__setattr__(name, value)


    def read(self, iprot):
        if iprot._fast_decode is not None and isinstance(iprot.trans, TTransport.CReadableTransport) and self.thrift_spec is not None:
            iprot._fast_decode(self, iprot, [self.__class__, self.thrift_spec])
            return
        iprot.readStructBegin()
        while True:
            (fname, ftype, fid) = iprot.readFieldBegin()
            if ftype == TType.STOP:
                break
            if fid == 1:
                if ftype == TType.STRING:
                    self.param1 = iprot.readString().decode('utf-8', errors='replace') if sys.version_info[0] == 2 else iprot.readString()
                else:
                    iprot.skip(ftype)
            elif fid == 2:
                if ftype == TType.I32:
                    self.param2 = TestEnum(iprot.readI32())
                else:
                    iprot.skip(ftype)
            elif fid == 3:
                if ftype == TType.I32:
                    self.param3 = shared_types.ttypes.SharedEnum(iprot.readI32())
                else:
                    iprot.skip(ftype)
            else:
                iprot.skip(ftype)
            iprot.readFieldEnd()
        iprot.readStructEnd()

    def write(self, oprot):
        if oprot._fast_encode is not None and self.thrift_spec is not None:
            oprot.trans.write(oprot._fast_encode(self, [self.__class__, self.thrift_spec]))
            return
        oprot.writeStructBegin('TestStruct')
        if self.param1 is not None:
            oprot.writeFieldBegin('param1', TType.STRING, 1)
            oprot.writeString(self.param1.encode('utf-8') if sys.version_info[0] == 2 else self.param1)
            oprot.writeFieldEnd()
        if self.param2 is not None:
            oprot.writeFieldBegin('param2', TType.I32, 2)
            oprot.writeI32(self.param2.value)
            oprot.writeFieldEnd()
        if self.param3 is not None:
            oprot.writeFieldBegin('param3', TType.I32, 3)
            oprot.writeI32(self.param3.value)
            oprot.writeFieldEnd()
        oprot.writeFieldStop()
        oprot.writeStructEnd()

    def validate(self):
        return

    def __repr__(self):
        L = ['%s=%r' % (key, value)
             for key, value in self.__dict__.items()]
        return '%s(%s)' % (self.__class__.__name__, ', '.join(L))

    def __eq__(self, other):
        return isinstance(other, self.__class__) and self.__dict__ == other.__dict__

    def __ne__(self, other):
        return not (self == other)
all_structs.append(TestStruct)
TestStruct.thrift_spec = (
    None,  # 0
    (1, TType.STRING, 'param1', 'UTF8', None, ),  # 1
    (2, TType.I32, 'param2', None, None, ),  # 2
    (3, TType.I32, 'param3', None, None, ),  # 3
)
fix_spec(all_structs)
del all_structs
```

**Comment by fishy** [comment]
actually I just remembered another corner case, we also override `__setattr__` on all exceptions, so we need to make sure that an exception with an enum field still works. please also add a test for that.

(you can check https://issues.apache.org/jira/browse/THRIFT-5715 & https://issues.apache.org/jira/browse/THRIFT-4002 for more context)

**Comment by roshanjrajan-zip** [comment]
@fishy Updated the PR with handling the Immutable struct logic and added to the test case.

File: compiler/cpp/src/thrift/generate/t_py_generator.cc [hunk 2]
```diff
@@ -871,7 +871,12 @@ void t_py_generator::generate_py_struct_definition(ostream& out,
       }
 
       if (is_immutable(tstruct)) {
-        if (gen_newstyle_ || gen_dynamic_) {
+        if (gen_enum_ && type->is_enum()) {
+          indent(out) << "super(" << tstruct->get_name() << ", self).__setattr__('"
```
Comment (fishy) [inline compiler/cpp/src/thrift/generate/t_py_generator.cc:875]: this is probably too much magic to my preference, makes me start to prefer the approach in #2824 again 😅

File: test/py/explicit_module/runtest.sh [hunk 3]
```diff
@@ -25,10 +25,12 @@ rm -rf gen-py
 ../../../compiler/cpp/thrift --gen py test3.thrift && exit 1  # Fail since test3.thrift has python keywords
 ../../../compiler/cpp/thrift --gen py:enum shared_types.thrift || exit 1
 ../../../compiler/cpp/thrift --gen py:enum test4.thrift || exit 1
+../../../compiler/cpp/thrift --gen py:enum test5.thrift || exit 1
```
Comment (fishy) [inline test/py/explicit_module/runtest.sh:28]: can you please also generate the code with both enum and slots (as slots also change the code significantly), you can use `--out` to put it under a different directory than `gen-py` so you can have another test around line 33 to also make sure that the combination of enum+slots still works.

File: compiler/cpp/src/thrift/generate/t_py_generator.cc (hunk 2 above)
Comment (roshanjrajan-zip) [reply compiler/cpp/src/thrift/generate/t_py_generator.cc:875]: Yeah I was thinking about this and then I realized that the fix was actually broken because I wasn't running one of the tests that I added.... 😓 . So actually this fix is needed to make sure existing users aren't broken.

```python
if __name__ == "__main__":
    serialization_deserialization_enum_test()
    serialization_deserialization_string_test
```

File: test/py/explicit_module/runtest.sh (hunk 3 above)
Comment (roshanjrajan-zip) [reply test/py/explicit_module/runtest.sh:28]: I am not sure what slots are and how I can generate code that has enum and slots. Is there an example how I can generate code with enum and slots? I tried variations of `py:enum py:slots` and such but no such luck.

Comment (roshanjrajan-zip) [reply test/py/explicit_module/runtest.sh:28]: NVM I figured it out and it seems to work. Is there a specific test case you have in mind? I just changed this - `../../../compiler/cpp/thrift --gen py:enum,slots test5.thrift` and the new serialization/deserialization test  that I added works.

File: compiler/cpp/src/thrift/generate/t_py_generator.cc (hunk 2 above)
Comment (fishy) [reply compiler/cpp/src/thrift/generate/t_py_generator.cc:875]: so this code:
```py

    def __init__(self, whatOp=None, why=None, who=None,):
        super(TestException, self).__setattr__('whatOp', whatOp)
        super(TestException, self).__setattr__('why', why if hasattr(why, 'value') else shared_types.ttypes.SharedEnum.__members__.get(why))
        super(TestException, self).__setattr__('who', who if hasattr(who, 'value') else TestEnum.__members__.get(who))
```
calls the parent's `__setattr__`, doesn't that essentially set those fields on the parent instead? I guess it still "works" as when accessing them you just inherit from the parent, but this _feels_ wrong to me.

File: test/py/explicit_module/runtest.sh (hunk 3 above)
Comment (fishy) [reply test/py/explicit_module/runtest.sh:28]: yea can you add that into the PR so we make sure neither `enum` and `enum,slots` will break in the future?

File: compiler/cpp/src/thrift/generate/t_py_generator.cc (hunk 2 above)
Comment (roshanjrajan-zip) [reply compiler/cpp/src/thrift/generate/t_py_generator.cc:875]: The code that generates `super(TestException, self).__setattr__('whatOp', whatOp)` already exists before my change and I am extending this pattern. In python, [super](https://docs.python.org/3/library/functions.html#super) is implemented as part of the binding process for explicit dotted attribute lookups such as __setattr__ and such. So I don't think this is a concern because this is how Python does things under the hood.

File: test/py/explicit_module/runtest.sh (hunk 3 above)
Comment (roshanjrajan-zip) [reply test/py/explicit_module/runtest.sh]: Added the slots option to the existing test5 test!

**Review by fishy — APPROVED** [review]
it's holiday/long weekend here, I'll probably merge this later next week if no one objects.
